Handle all enum values in switch/case [blocks: #2310]

tautschnig opened the pull request.

tautschnig committed:
Handle all enum values in switch/case

smowton commented:
What crazy compiler won't let you use `default:`?

tautschnig commented:
> What crazy compiler won't let you use `default:`?

I'd say that's a perfectly sane compiler as far as enums are concerned: it helps you not forget about values that were added to the enum type later on, but aren't handled in `switch` statements. But I do accept that I might be the only one thinking this is a better approach.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: ca1ceb95a5fcba43597dee515159761f6c474f20).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78386231

smowton commented:
your branch name has a `vs-` prefix though, does that imply MSVC was complaining?

tautschnig commented:
> [...] does that imply MSVC was complaining?

Yes, Visual Studio does complain about cases not being handled (despite the `default`).

smowton commented on src/ansi-c/c_typecast.cpp:
```diff
@@ -396,8 +396,13 @@ void c_typecastt::implicit_typecast_arithmetic(
+  case FIXEDBV:
+  case LARGE_UNSIGNED_INT:
+  case LARGE_SIGNED_INT:
+                   return; // do nothing
```
Weird indent

smowton reviewed (approved):
Fair point, in that case I support the change. If you do this, consider enabling -Wswitch-enum in our GCC configs so any mistake is detected early, rather than presenting developers with an annoying stumbling block when the shorthand is only flagged on Windows.

tautschnig committed:
Add -Wswitch-enum to default GCC/Clang build options

This is now consistent with the warnings that Visual Studio would generate,
which warns about missing enum cases in switch/case even when a default: is
present.

chrisr-diffblue commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -496,6 +496,7 @@ static mp_integer max_value(const typet &type)
   else if(type.id() == ID_unsignedbv)
     return to_unsignedbv_type(type).largest();
   UNREACHABLE;
+  return 0;
```
It seems like this change (and the others in this commit) are somewhat independent of the stated purpose of this  PR or are these warnings triggered by the change in switch/case statements?

tautschnig replied:
I'm afraid they are: making Visual Studio happy that all cases are covered implies that the `default` can be removed. Doing so makes GCC unhappy as it thinks code becomes reachable. The specific code that you highlighted above, though, is not linked to one of those `default`-is-gone cases, but did fit the same topic. Let me know if you'd want me to dissect it.

chrisr-diffblue replied:
Thanks for the clarification - I'm fine with this PR as it is :-)

chrisr-diffblue reviewed (approved):
Generally approve of the change. A minor nit being a possible independent commit being included in this PR, but given the PR is one of a series of nice clean ups related to compiler warnings, I'm not going to block this PR :-)

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 04f3ac309f5283617dd9165191c496a683ae21d8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78602309
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

peterschrammel commented on src/analyses/custom_bitvector_analysis.cpp:
```diff
@@ -536,7 +536,20 @@ void custom_bitvector_domaint::transform(
+  case INCOMPLETE_GOTO:
+  case NO_INSTRUCTION_TYPE:
     {
     }
```
I'd prefer to have a `break` here to avoid stupid bugs when someone adds a new case at the end (several occurrences).

martin-cs replied:
At the risk of saying #751, I think there is some subtlety that is being missed here (and probably should have always been documented, etc.)  Some of these are ignored as it is the correct transform (ASSERT, SKIP), some are ignored because it is a valid over-approximation (ASSUME), some are ignored because the domain wasn't really intended to be used with that kind of program (ATOMIC_* *_THREAD) and it's not really sound but maybe it's OK, some are assumed to be removed by previous passes (THROW, CATCH, RETURN) and so are indicative of a deeper problem.

If we're going to add all of these, as part of this PR or after, I think we should actually document which is which and add INVARIANTs etc.

peterschrammel commented on src/goto-analyzer/taint_analysis.cpp:
```diff
@@ -68,143 +68,135 @@ void taint_analysist::instrument(
+        identifiers.insert(identifier);
 
-          identifiers.insert(identifier);
+        irep_idt class_id=function.get(ID_C_class);
```
`const &`?

peterschrammel commented on src/goto-analyzer/taint_analysis.cpp:
```diff
@@ -68,143 +68,135 @@ void taint_analysist::instrument(
+    if(instruction.is_function_call())
     {
-    case FUNCTION_CALL:
+      const code_function_callt &function_call=
```
I assume clang-format will complain here and in other places...

peterschrammel commented on src/goto-programs/vcd_goto_trace.cpp:
```diff
@@ -109,40 +109,32 @@ void output_vcd(
+    if(step.is_assignment())
     {
-    case goto_trace_stept::typet::ASSIGNMENT:
+      irep_idt identifier=step.lhs_object.get_identifier();
```
`const &`?

peterschrammel commented on src/goto-symex/symex_main.cpp:
```diff
@@ -463,7 +463,7 @@ void goto_symext::symex_step(
-  default:
-    throw "symex got unexpected instruction";
+  case INCOMPLETE_GOTO:
+    throw "symex got INCOMPLETE_GOTO";
```
`INVARIANT`?

peterschrammel commented on src/config.inc:
```diff
@@ -5,7 +5,7 @@ BUILD_ENV = AUTO
   #CXXFLAGS += /Wall /WX
 else
-  CXXFLAGS += -Wall -pedantic -Werror -Wno-deprecated-declarations
+  CXXFLAGS += -Wall -pedantic -Werror -Wno-deprecated-declarations -Wswitch-enum
```
And here too: https://github.com/diffblue/cbmc/blob/develop/CMakeLists.txt#L26

martin-cs commented on src/solvers/smt2/smt2_dec.cpp:
```diff
@@ -140,7 +140,7 @@ decision_proceduret::resultt smt2_dect::dec_solve()
 
-  default:
+  case solvert::GENERIC:
     assert(false);
```
If we are changing the default behaviour, can we do something better than just assert(false)?

martin-cs reviewed (changes requested):
Can we have the corresponding change to the coding standard and something (compiler flag) to enforce this in the future?

tautschnig commented:
@martin-cs:
> Can we have the corresponding change to the coding standard and something (compiler flag) to enforce this in the future?

The compiler flag is in place. I'd prefer to go with this and force people to get it right rather than having them read docs :-P

tautschnig committed:
Disable clang-format for recently touched switch statements

We should not re-indent a large number of lines of code just because some case
statements were added. Silence clang-format for these instead.

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 54b561db89fd810fa67d5fad2f5637a1b8937dfa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90737128

peterschrammel commented on jbmc/src/miniz/miniz.cpp:
```diff
@@ -7194,3 +7198,4 @@ mz_bool mz_zip_end(mz_zip_archive *pZip)
 #ifdef __cplusplus
 }
 #endif
+// clang-format off
```
🚫 on?

tautschnig replied:
Oops, fixed, thank you!

peterschrammel commented on src/analyses/interval_domain.cpp:
```diff
@@ -64,6 +64,7 @@ void interval_domaint::transform(
   ai_baset &,
   const namespacet &ns)
 {
+  // clang-format off
```
This seems excessive use of `clang-format off` (several occurrences). May I ask what undesirable formatting clang-format produces in this case?

tautschnig replied:
clang-format is absolutely doing the right thing here, but I wasn't sure I should be re-formatting several hundred lines of code (some of these inconsistently formatted switch/case statements span a lot of code). I would be easy enough for me to ditch this commit (those `clang-format off/on` are all in a separate commit) and apply proper formatting instead.

peterschrammel replied:
I have a preference for reformatting it properly right away. Turning off the formatter over hundreds of lines will not make it better in future.

tautschnig replied:
Done.

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 7ad2c8eddce69523de11aedc75e8412a1089ee9a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90766520

tautschnig committed:
clang-format switch statements touched in this PR

tautschnig commented:
I have added the reformatting patch as suggested. Please strictly review commit-by-commit, the last (and largest) one is whitespace-only.

allredj reviewed (commented):
:warning:
Diffblue compatibility check is currently unavailable.
Please create manual bump.
(cbmc commit: 3525d611bd7186beb72eb8b7bbbc4c17a2f8d802).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90838898

allredj commented:
That PR passes the compatibility checks fine! :heavy_check_mark:

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 07edcf6e633a90182637c7a45a1c817c3aa9ea79).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90995517

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 81f679ada3d44b67433a03385b34a29781319426).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91023424

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b62745468bd64571055155bbdaa2fc09a782614d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91029723

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fb62662f91970b9655df5da76d49d1dea5691b86).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95498796

tautschnig commented:
@martin-cs Would you mind taking another look?

Commit:
Annotate missed cases in the transform functions of various domains

There are a number of reasons why instruction types were left out of
these case statements :

1. Ignoring this instruction is generally a valid overapproximation.
2. Ignoring this instruction is a valid overapproximation for this
   domain.
3. The instruction is assumed to not be present due to preceding
   passes.
4. The instruction should never appear in any valid goto program.
5. The instruction is newer than the analysis code and was forgotten.

This patch tries to correctly document which of these apply.

martin-cs reviewed (commented):
My first concern (about enforcement) has been resolved.  My second (about hiding errors) should be in part addressed by this patch : https://github.com/martin-cs/cbmc/tree/vs-switch-case-with-checks or something like it.

martin-cs reviewed (approved):
Thanks for merging that; I hope it hasn't broken everything!

tautschnig committed:
Disable a number of invariants that currently fail

This takes us back to the behaviour prior to this series of commits, and
effectively is a to-do list to be addressed. We should either handle the cases,
or get rid of the instruction type.

martin-cs reviewed (approved):
I wish I could say I was surprised by this.  See also https://github.com/diffblue/cbmc/issues/4323 for a discussion of what to do about this.

tautschnig commented:
@kroening @peterschrammel Please review commit-by-commit, where you may wish to skip the clang-format-only commit.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7b51b0c554e700e26b5a9453bd5e2bac83e8a48e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103086699
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

peterschrammel commented on src/analyses/uncaught_exceptions_analysis.cpp:
```diff
@@ -125,23 +125,30 @@ void uncaught_exceptions_domaint::transform(
+    break;
   case INCOMPLETE_GOTO:
   case NO_INSTRUCTION_TYPE:
+    DATA_INVARIANT(false, "Only complete instructions can be analysed");
```
:pick: analyzed (several occurrences)

tautschnig replied:
All done.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0727459626266dab00ea8055d14b9537d7862aac).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103501248

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: bf2df045956cf4a304453ff561fb24e8cf84b5cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104505124
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
